lib/python/qmk/cli/new/keyboard.py

```diff
@@ -1,11 +1,133 @@
-"""This script automates the creation of keyboards.
+"""This script automates the creation of new keyboard directories using a starter template.
 """
+from datetime import date
+import fileinput
+from pathlib import Path
+import re
+import shutil
+
+from qmk.commands import git_get_username
+import qmk.path
 from milc import cli
+from milc.questions import choice, question
+
+KEYBOARD_TYPES = ['avr', 'ps2avrgb']
+
+
+def keyboard_name(name):
+    """Callable for argparse validation.
+    """
+    if not validate_keyboard_name(name):
+        raise ValueError
+    return name
+
+
+def validate_keyboard_name(name):
+    """Returns True if the given keyboard name contains only lowercase a-z, 0-9 and underscore characters.
+    """
+    regex = re.compile(r'^[a-z0-9/_]+$')
+    return bool(regex.match(name))
 
 
-@cli.subcommand('Creates a new keyboard')
+@cli.argument('-kb', '--keyboard', help='Specify the name for the new keyboard directory', arg_only=True, type=keyboard_name)
+@cli.argument('-t', '--type', help='Specify the keyboard type', arg_only=True, choices=KEYBOARD_TYPES)
+@cli.argument('-u', '--username', help='Specify your username (default from Git config)', arg_only=True)
+@cli.subcommand('Creates a new keyboard directory')
 def new_keyboard(cli):
-    """Creates a new keyboard
+    """Creates a new keyboard.
+    """
+    cli.log.info('{style_bright}Generating a new QMK keyboard directory{style_normal}')
+    cli.echo('')
+
+    # Get keyboard name
+    new_keyboard_name = None
+    while not new_keyboard_name:
+        new_keyboard_name = cli.args.keyboard if cli.args.keyboard else question('Keyboard Name:')
+        if not validate_keyboard_name(new_keyboard_name):
+            cli.log.error('Keyboard names must contain only {fg_cyan}lowercase a-z{fg_reset}, {fg_cyan}0-9{fg_reset}, and {fg_cyan}_{fg_reset}! Please choose a different name.')
+
+            # Exit if passed by arg
+            if cli.args.keyboard:
+                return False
+
+            new_keyboard_name = None
+            continue
+
+        keyboard_path = qmk.path.keyboard(new_keyboard_name)
+        if keyboard_path.exists():
+            cli.log.error(f'Keyboard {{fg_cyan}}{new_keyboard_name}{{fg_reset}} already exists! Please choose a different name.')
+
+            # Exit if passed by arg
+            if cli.args.keyboard:
+                return False
+
+            new_keyboard_name = None
+
+    # Get keyboard type
+    keyboard_type = cli.args.type if cli.args.type else choice('Keyboard Type:', KEYBOARD_TYPES, default=0)
+
+    # Get username
+    user_name = None
+    while not user_name:
+        git_username = git_get_username()
```

Comment:
I think I'd have git username as a fallback if `cli.config.user.name` is not defined.

Reply:
This is just keeping the same behaviour as the existing script (as much as possible, anyway).

Reply:
I think that's mostly because when the script was created, `user.name` did not exist.
I don't think checking the git username is a bad idea, but preferring `user.name` over it sounds like a no-brainer.

Reply:
Yeah, I think I misunderstood what you meant. This should do the trick, I think.